lib/SIL/GraphFunctionDeviceInfo.cpp

```diff
@@ -184,6 +184,47 @@ bool GraphFunctionDeviceInfo::isConfigOp(const GraphOperationInfo &opInfo) {
          opInfo.getOperationName() == "tfc.configureCPU";
 }
 
+void GraphFunctionDeviceInfo::finalizeUsedDevices() {
+  // In this edge case, all ops are placed on the ALL device (e.g. Const). In
+  // that case, usedDeviceIds should contain the RUNTIME device.
+  if (usedDeviceIds.empty()) {
+    primaryDeviceId = RuntimeDeviceId;
+    usedDeviceIds.insert(RuntimeDeviceId);
+    return;
+  }
+
+  // For device partitioning to work, the device set cannot include the RUNTIME
+  // device along with some other device(s). This is because we don't know what
+  // the RUNTIME device is at compile time, so we cannot create sends/recvs
+  // graph nodes.
+  if (usedDeviceIds.size() > 1 && usedDeviceIds.count(RuntimeDeviceId))
+    assert(0 && "Cannot yet handle a multi-device function involving the "
+                "RUNTIME device");
+
+  // SIL functions can be processed in non-deterministic ordering, so the
+  // ordering of device ids being inserted into `deviceInfo` is not
+  // deterministic either.
+  // To make sure we produced deterministic SIL code (e.g. produce graph
+  // function for CPU, before GPU), which is useful at least for unit testing,
+  // we sort the device IDs.
+  SmallVector<DeviceId, 8> deviceIds(usedDeviceIds.begin(),
+                                     usedDeviceIds.end());
+  assert(!deviceIds.empty());
+  llvm::array_pod_sort(deviceIds.begin(), deviceIds.end());
+  usedDeviceIds.clear();
+  usedDeviceIds.insert(deviceIds.begin(), deviceIds.end());
```

Comment:
Indeed it's quite confusing. 

`SmallSet` (different from std::set) could be backed by a SmallVector instead of a tree-based container, so we need to sort its elements. Added a comment to clarity this.